### PR TITLE
fix: Override blockhash in zk context

### PR DESCRIPTION
**Pull request by Jrigada**
## Motivation
blockhash was different in non-zk execution than the one on zk execution when executing tests. 

## Solution
Override blockhash in zkcontext

### CODE REVIEW

File: crates/zksync/core/src/vm/tracer.rs [hunk 1]
```diff
@@ -286,6 +292,20 @@ impl<S: Send, H: HistoryMode> DynTracer<S, SimpleMemory<H>> for CheatcodeTracer
...
+        // Override blockhash
+        if let Opcode::FarCall(_call) = data.opcode.variant.opcode {
+            let calldata = get_calldata(&state, memory);
+            let current = state.vm_local_state.callstack.current;
+            let zero: [u8; 32] = [0; 32];
+
+            if current.code_address == SYSTEM_CONTEXT_ADDRESS &&
+                calldata.starts_with(&SELECTOR_BLOCK_HASH)
+            {
+                self.farcall_handler.set_immediate_return(zero.to_vec());
```
Comment (nbaztec) [inline crates/zksync/core/src/vm/tracer.rs:304]:
```suggestion
                self.farcall_handler.set_immediate_return(rU256::ZERO.to_be_bytes_vec());
```

File: zk-tests/src/Cheatcodes.t.sol [hunk 2]
```diff
@@ -190,4 +191,10 @@ contract ZkCheatcodesTest is Test {
 
         transactor.transact();
     }
+
+    function testBlockHash() public {
+        BlockhashCaller caller = new BlockhashCaller();
+        bytes32 hash = caller.getBlockhash(1);
+        require(hash == bytes32(0), "blockhash mismatch");
+    }
```
Comment (nbaztec) [inline zk-tests/src/Cheatcodes.t.sol:199]: This test can be part of `testZkPropagatedBlockEnv` in `Basic.t.sol` where the blocknumber can be hardcoded. We should check that the blockhash is zero both outside and within the zksync context.

**Review by Karrq — CHANGES_REQUESTED** [review]
When we looked at the era-contracts implementation there was a comment indicating that in EVM, a 0 is returned for blockhash when the block is older than 256. Could this mean that newer blocks will return a different hash?

Could you add a test that captures this behavior also? Using fork we can get a proper chain and retrieve the last block's hash using `block.number`

File: crates/cheatcodes/src/inspector.rs [hunk 3]
```diff
@@ -1738,13 +1738,20 @@ impl<DB: DatabaseExt + Send> Inspector<DB> for Cheatcodes {
...
             ) {
-                self.combined_logs.extend(result.logs.into_iter().map(|log| {
+                self.combined_logs.extend(result.logs.clone().into_iter().map(|log| {
                     Some(Log {
                         address: log.address,
                         data: LogData::new_unchecked(log.topics, log.data),
                     })
                 }));
 
+                //for each log in cloned logs call handle_expect_emit
```
Comment (nbaztec) [inline crates/cheatcodes/src/inspector.rs:1748]:
```suggestion
                // for each log in cloned logs call handle_expect_emit
```

File: zk-tests/src/Cheatcodes.t.sol [hunk 4]
```diff
@@ -45,7 +46,22 @@ contract MyProxyCaller {
     }
 }
 
+contract Emitter {
```
Comment (Karrq) [inline zk-tests/src/Cheatcodes.t.sol:49]: Do these changes belong to this PR?

**Review by Karrq — CHANGES_REQUESTED** [review]

**Review by Karrq — APPROVED** [review]

**Pull request merged**